### ISSUE TITLE
Considerations on computing the calibration curve

### ISSUE THREAD

**Issue by giacomoran** [open-spaced-repetition/spaced-repetition-algorithm-metric#2]
I'll share some more thoughts on calibration curves and $ICI$, continuing [our discussion from x.com](https://x.com/JarrettYe/status/1746708553593974933?s=20).

In discussing the computation of $ICI$, the question arised of whether Python's `lowess` function can replace R's `loess`, `lowess` functions. I'll try to frame the problem more broadly. 

### Problem

The problem the computation of the calibration curve: we fit a curve on the $Y$ binary observations (y-axis) against $\hat{P}$ predicted probabilities (x-axis), the result is a smooth curve of $\hat{P}_c$ calibrated predicted probabilities (y-axis) against $\hat{P}$ predicted probabilities (x-axis).

Once we obtain the calibration curve, we can:
- plot the calibration curve (ideally along a plot of the empirical distribution of the predicted probabilities)
- compute $ICI$, $E_{max}$ and similar indices, as described in the paper by Austin and Steyerberg

### Solutions

`loess`, `lowess` and similar functions are different smoothing techniques, to obtain $\hat{P}_c$ from $Y$. One could also use splines.
Regarding the current binning method currently used in `metrics_research.ipynb` to plot the calibration curves, I argue that it is one more smoothing technique.

The issue is picking a smoothing technique among the many possible alternatives.

Here is a plot of different calibration curves fitted on the data from `metrics_research.ipynb` (the `lowess` function is the one from Python; I've kept the plot of empirical distribution of predicted probabilities computed by binning, it could be replaced by a different method).

![calibration](https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/assets/8596184/ccd2bca1-cd88-4990-9aa6-08f3986ddcf0)

R's `loess`, `lowess` and Python's `lowess` are different implementations of the same idea. They use slightly different algorithms and parameters, but I don't think differences matter much.

Pros/Cons Using Python's `lowess` function instead of binning:
- (+) less degrees of freedom / avoid having to pick a binning strategy
- (+) no need for manual implementation of binning
- (-) less popular in machine learning literature

### Suggestions

My suggestions is to pick a single smoothing technique for computing the calibration curve, and to stick to it: use it both for plotting the calibration curve and for computing $ICI$ and similar indices.
You might want to explore the current literature for more appropriate smoothing techniques, my efforts in doing so stopped 2 years ago.

### References

- Austin, P., & Steyerberg, E. (2019). The Integrated Calibration Index (ICI) and related metrics for quantifying the calibration of logistic regression models. Statistics in Medicine, 38. https://doi.org/10.1002/sim.8281
- [Documentation for Python's `lowess` function](https://www.statsmodels.org/devel/generated/statsmodels.nonparametric.smoothers_lowess.lowess.html)
- [R's `loess` vs `lowess`](https://stats.stackexchange.com/a/472370)
- [Wikipedia: Smoothing](https://en.wikipedia.org/wiki/Smoothing)

**Comment by L-M-Sherlock** [comment]
Yeah. And I find that lowess is faster than loess in R. It's more simple, too.

**Comment by L-M-Sherlock** [comment]
I implement a loss function based on ICI:

```python
from torch import Tensor, nn
import torch
from statsmodels.nonparametric.smoothers_lowess import lowess


class ICILoss(nn.Module):
    def __init__(self):
        super(ICILoss, self).__init__()

    def forward(self, predictions: Tensor, labels: Tensor) -> Tensor:
        observation = lowess(
            labels.detach().numpy(),
            predictions.detach().numpy(),
            it=0,
            delta=0.01
            * (max(predictions.detach().numpy()) - min(predictions.detach().numpy())),
            return_sorted=False,
        )
        observation = torch.as_tensor(observation, dtype=torch.float32)
        return torch.abs(predictions - observation)
```

**Comment by giacomoran** [comment]
What do you plan on using it for?

**Comment by L-M-Sherlock** [comment]
I test it with FSRS in https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/blob/main/ICILoss.ipynb

It really reduces the ICI. However, I find a method to cheat ICI: https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/blob/main/cheat_ICI.ipynb

**Comment by L-M-Sherlock** [comment]
@Expertium you're likely to be interested in it. I implement a loss function that could reduce RMSE(bins) and ICI significantly. But I'm afraid that it just finds an effective way to cheat the metrics.

**Comment by Expertium** [comment]
I'll benchmark it to see how it improves RMSE (bins) and logloss

**Comment by L-M-Sherlock** [comment]
Don't forget to update the hyper-parameters.

**Comment by Expertium** [comment]
What do you mean?

**Comment by Expertium** [comment]
@L-M-Sherlock 
```
C:\Users\Andrew\AppData\Local\Programs\Python\Python310\lib\site-packages\statsmodels\nonparametric\smoothers_lowess.py:227: RuntimeWarning: invalid value encountered in divide
  res, _ = _lowess(y, x, x, np.ones_like(x),
```

**Comment by L-M-Sherlock** [comment]
> What do you mean?

I mean the ICILoss only work with large batch size and low learning rate. You check the modified hyper-parameters in the notebook.

**Comment by Expertium** [comment]
Oh well, I'm already running it. Alright, I'll rerun it again later. At the very least, we'll get to see how it will perform without changing anything else.

**Comment by Expertium** [comment]
I finished running the benchmark without touching hyperparameters. The result is about the same as with the logloss, RMSE decreased by a fraction of a percentage point. While it's statistically significant, it's practically useless. Now I'm benchmarking with hyperparameters recommended by LMSherlock.

**Comment by giacomoran** [comment]
> I test it with FSRS in https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/blob/main/ICILoss.ipynb

I'm very surprised that Torch is able to differentiate through `lowess`.. interesting, I'm curious about the results.

> It really reduces the ICI. However, I find a method to cheat ICI: https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/blob/main/cheat_ICI.ipynb

Help me understand the `cheat_ICI` notebook. To summarise what you are doing, you are computing a new set of predicted probabilities (here I call it `p_smoothed`) by smoothing the true labels `y` sorted by `p`:

```
p = # predicted probabilities
y = # true labels

# do: sort p and y by p

p_smoothed = np.convolve(y, np.ones(10), 'same') / 10 # smooth y
```

From the results, `ICI` and related metrics decrease while `logLoss` increases.

Is that correct? Are you saying it's cheating because `logLoss` increases while `ICI` decreases?
Since you are using the true labels `y`, the cheating applies only on a training set, not on a validation or test set, correct?
I haven't tried it, but my intuition is that you can cheat RMSE(bins) in the same way.

**Comment by L-M-Sherlock** [comment]
> I'm very surprised that Torch is able to differentiate through `lowess`.. interesting,

Torch doesn't differentiate through `lowess` because I just use `lowess` to smoothen the labels.

> Is that correct? Are you saying it's cheating because `logLoss` increases while `ICI` decreases?

Yeah, and it even outperforms the prediction of real probability.

**Comment by Expertium** [comment]
I finished benchmarking with recommended hyperparameters. The results are actually worse. I couldn't use WIlcoxon's test because the number of collections is somehow not exactly the same, don't ask how. Maybe it has something to do with time splits.
Anyway, here are the results:

![image](https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/assets/83031600/2a0dc054-5b38-492b-aa87-dbeebef01d43)

I will run one mroe benchmark, with the following loss function:
```
    def forward(self, predictions: Tensor, labels: Tensor) -> Tensor:
        observation = lowess(
            labels.detach().numpy(),
            predictions.detach().numpy(),
            it=0,
            delta=0.01
            * (max(predictions.detach().numpy()) - min(predictions.detach().numpy())),
            return_sorted=False,
        )
        observation = torch.as_tensor(observation, dtype=torch.float32)
        observation = observation.clamp(0.0001, 0.9999)
        wozloss = -torch.log(1 - torch.abs(predictions - observation))
        return wozloss.clamp(0, 100)
```
I've talked about this before, but the problem with logloss is that it behaves extremely unituitively when both predictions and labels are continuous. So I use Woz's formula instead, because Wozloss(p, p) = 0, but logloss(p, p) != 0 if p is not binary, which is a very undesirable property.

> you can cheat RMSE(bins) in the same way.

RMSE (bins) can be cheated even without any smoothing. Suppose the average retention of the user is 90%. The algorithm always outputs "probability of recall = 90%" no matter what; therefore, all algorithmic predictions will fall into only one bin, and other bins will be empty. Within that bin, the average predicted probability of recall is obviously 90%, and the actual retention is 90%. That would result in RMSE = 0%.

**Comment by L-M-Sherlock** [comment]
> I finished benchmarkign with recommended hyperparameters. The results are actually worse.

That's weird. I tested it in the collections of you, me and user1839. The results were all pretty good. Maybe it only works on large collections? Maybe I need to do more research here. I have pushed the ICILoss branch in FSRS-optimizer repo. You can test it on your collection or decks.

**Comment by Expertium** [comment]
I finished benchmarking the idea with wozloss above. It slightly increased logloss and slightly decreased RMSE, nothing important.
![image](https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric/assets/83031600/eb8e1991-910e-4a5c-b8d7-5cc2c7c32a38)
